File: src/scala/implementation/InnerSet_4.scala
package scala.implementation


class InnerSet_4[T](protected val element1: T, protected val element2: T, protected val element3: T, protected val element4: T) extends AbstractInnerSet[T]
{

    def this(p_innerSet: InnerSet_4[T]) = this(p_innerSet.element1 , p_innerSet.element2 , p_innerSet.element3 , p_innerSet.element4 )
    override def addElement(p_element: T): InnerSet[T] =
    {
        if (element1.equals(p_element) || element2.equals(p_element) || element3.equals(p_element) || element4.equals(p_element))
        {
            this
        }
        else  new InnerSet_5(element1, element2, element3, element4, p_element)
    }

    override def addUnchecked(p_element: T): InnerSet[T] =  new InnerSet_5(element1, element2, element3, element4, p_element)

    override def addAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =  p_innerSet.addElement(element1).addElement(element2).addElement(element3).addElement(element4)

    override def containsElement(p_element: Object): Boolean = element1.equals(p_element) || element2.equals(p_element) || element3.equals(p_element) || element4.equals(p_element)

    override def containsAllElements(p_innerSet: InnerSet[T]): Boolean =  p_innerSet.containsElement(element1.asInstanceOf[Object]) && p_innerSet.containsElement(element2.asInstanceOf[Object]) && p_innerSet.containsElement(element3.asInstanceOf[Object]) && p_innerSet.containsElement(element4.asInstanceOf[Object])

    override def getElement(p_index: Int): Option[T] = p_index match{
        case 1 => Some(element1)
        case 2 => Some(element2)
        case 3 => Some(element3)
        case 4 => Some(element4)
        case _ => None
    }


    override def removeElement(p_element: Object): InnerSet[T] =
    {
            if(element1.equals(p_element))
         new InnerSet_3(element2, element3, element4)
                  else             if(element2.equals(p_element))
         new InnerSet_3(element1, element3, element4)
                  else             if(element3.equals(p_element))
         new InnerSet_3(element1, element2, element4)
                  else             if(element4.equals(p_element))
         new InnerSet_3(element1, element2, element3)
                        else this
    }

    override def removeAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =  p_innerSet.removeElement(element1.asInstanceOf[Object]).removeElement(element2.asInstanceOf[Object]).removeElement(element3.asInstanceOf[Object]).removeElement(element4.asInstanceOf[Object])

    override def retainAllElements(p_innerSet: InnerSet[T]): InnerSet[T] =
    {
      var i: InnerSet[T] = this
            if(!p_innerSet.containsElement(element1.asInstanceOf[Object])) i = i.removeElement(element1.asInstanceOf[Object])
            if(!p_innerSet.containsElement(element2.asInstanceOf[Object])) i = i.removeElement(element2.asInstanceOf[Object])
            if(!p_innerSet.containsElement(element3.asInstanceOf[Object])) i = i.removeElement(element3.asInstanceOf[Object])
            if(!p_innerSet.containsElement(element4.asInstanceOf[Object])) i = i.removeElement(element4.asInstanceOf[Object])
            i
    }

    override def getSize: Int = 4

    override def iterator: InnerSetIterator[T] = new InnerSetIterator(this)

    override def copy: InnerSet[T] = this

    override def clear(unused: Boolean): InnerSet[T] = new InnerSet_0[T]

    override def toString : String = "InnerSet4 : { " + element1 + ", " + element2 + ", " + element3 + ", " + element4 + " }"

    override def addAndPropagate(p_element: T, p_microSet: MicroSet[T]): InnerSet[T] =
    {
        if (element1.equals(p_element) || element2.equals(p_element) || element3.equals(p_element) || element4.equals(p_element))
        {
            this
        }
        else         {
            p_microSet.add(p_element)
            new InnerSet_5(element1, element2, element3, element4, p_element)
        }    }

    override def addAllAndPropagate(p_innerSet: InnerSet[T], p_microSet: MicroSet[T]): InnerSet[T] = p_innerSet.addAllAndPropagateReverse(this, p_microSet)

    override def addAllAndPropagateReverse(p_innerSet: InnerSet[T], p_microSet: MicroSet[T]): InnerSet[T] =  p_innerSet.addAndPropagate(element1,p_microSet).addAndPropagate(element2,p_microSet).addAndPropagate(element3,p_microSet).addAndPropagate(element4,p_microSet)
}

File: src/cpp/implementation/inner_set_4.cpp
#include "inner_set_5.hpp"

#include "inner_set_4.hpp"

template<typename T>
inner_set_4<T>::inner_set_4()
{

}

template<typename T>
inner_set_4<T>::inner_set_4(T p_element_1, T p_element_2, T p_element_3, T p_element_4)
{
  a_values.element_1 = p_element_1;
  a_values.element_2 = p_element_2;
  a_values.element_3 = p_element_3;
  a_values.element_4 = p_element_4;
}

template<typename T>
inner_set_4<T>::inner_set_4(inner_set_3<T>* p_inner_set, T p_element)
{
  a_values.element_1 = p_inner_set->a_values.element_1;
  a_values.element_2 = p_inner_set->a_values.element_2;
  a_values.element_3 = p_inner_set->a_values.element_3;

  a_values.element_4 = p_element;

  delete p_inner_set;
}

template<typename T>
inner_set_4<T>::~inner_set_4()
{

}

template<typename T>
inner_set<T>* inner_set_4<T>::add_element(T p_element)
{
    if(!contains_element(p_element))
        return new inner_set_5<T>(this, p_element);
    else
        return this;
}

template<typename T>
inner_set<T>* inner_set_4<T>::add_all_elements(inner_set<T>* p_inner_set)
{
  return p_inner_set->add_element(a_values.element_1)->add_element(a_values.element_2)->add_element(a_values.element_3)->add_element(a_values.element_4);
}

template<typename T>
bool inner_set_4<T>::contains_element(T p_element)
{
  return  a_values.element_1 == p_element ||
  a_values.element_2 == p_element ||
  a_values.element_3 == p_element ||
  a_values.element_4 == p_element;
}

template<typename T>
bool inner_set_4<T>::contains_all_elements(inner_set<T>* p_inner_set)
{
  return p_inner_set->contains_element(a_values.element_1)
    && p_inner_set->contains_element(a_values.element_2)
    && p_inner_set->contains_element(a_values.element_3)
    && p_inner_set->contains_element(a_values.element_4);
}

template<typename T>
T inner_set_4<T>::get_element(int p_index)
{
  return reinterpret_cast<T*>(&a_values)[p_index];
}

template<typename T>
inner_set<T>* inner_set_4<T>::remove_element(T p_element)
{
  if(a_values.element_1 == p_element)
    return new inner_set_3<T>( a_values.element_2,  a_values.element_3,  a_values.element_4);
  else             if(a_values.element_2 == p_element)
    return new inner_set_3<T>( a_values.element_1,  a_values.element_3,  a_values.element_4);
  else             if(a_values.element_3 == p_element)
    return new inner_set_3<T>( a_values.element_1,  a_values.element_2,  a_values.element_4);
  else             if(a_values.element_4 == p_element)
    return new inner_set_3<T>( a_values.element_1,  a_values.element_2,  a_values.element_3);
  else
    return this;
}

template<typename T>
inner_set<T>* inner_set_4<T>::remove_all_elements(inner_set<T>* p_inner_set)
{
  return p_inner_set->remove_element(a_values.element_1)->remove_element(a_values.element_2)->remove_element(a_values.element_3)->remove_element(a_values.element_4);
}

template<typename T>
inner_set<T>* inner_set_4<T>::retain_all_elements(inner_set<T>* p_inner_set)
{
  inner_set<T>* r_inner_set = new inner_set_0<T>();

  if(p_inner_set->contains_element(a_values.element_1))
  {
    r_inner_set = r_inner_set->add_element(a_values.element_1);
  }

  if(p_inner_set->contains_element(a_values.element_2))
    r_inner_set = r_inner_set->add_element(a_values.element_2);

  if(p_inner_set->contains_element(a_values.element_3))
    r_inner_set = r_inner_set->add_element(a_values.element_3);

  if(p_inner_set->contains_element(a_values.element_4))
    r_inner_set = r_inner_set->add_element(a_values.element_4);

  return r_inner_set;
}

template<typename T>
int inner_set_4<T>::get_size()
{
    return 4;
}
